### PR TITLE
handle partial responses

### PR DESCRIPTION
**Pull request by almogdepaz**
send batch block requests
handle partial responses

### CODE REVIEW

File: sync/validation_queue.go [hunk 1]
```diff
@@ -74,25 +74,26 @@ func (vq *blockQueue) handleBlock(bjb fetchJob) {
 		mp[item.Hash32()] = tmp
 	}
```
Comment (almogdepaz) [inline sync/validation_queue.go:70]: this is a revert of #1409

File: sync/validation_queue.go [hunk 2]
```diff
@@ -74,25 +74,26 @@ func (vq *blockQueue) handleBlock(bjb fetchJob) {
...
-	block, found := mp[id]
-	if !found {
-		vq.updateDependencies(id, false)
-		vq.Error(fmt.Sprintf("could not retrieve a block in view "))
-		return
-	}
+		block, found := mp[id]
+		if !found {
+			vq.updateDependencies(id, false)
+			vq.Error(fmt.Sprintf("could not retrieve a block in view "))
```
Comment (zalmen) [inline sync/validation_queue.go]: print the block id

File: sync/validation_queue.go [hunk 3]
```diff
@@ -76,25 +76,26 @@ func (vq *blockQueue) handleBlock(bjb fetchJob) {
...
 
-	vq.Info("fetched  %v", id.String())
-	vq.visited[id] = struct{}{}
-	if err := vq.fastValidation(block); err != nil {
-		vq.Error("ValidationQueue: block validation failed", log.BlockId(uint64(block.ID())), log.Err(err))
-		vq.updateDependencies(id, false)
-		return
-	}
+		vq.Info("fetched  %v", id.String())
+		vq.visited[id] = struct{}{}
```
Comment (zalmen) [inline sync/validation_queue.go]: check if you really need `visited`

File: sync/validation_queue.go [hunk 4]
```diff
@@ -76,25 +76,26 @@ func (vq *blockQueue) handleBlock(bjb fetchJob) {
...
+		if err := vq.fastValidation(block); err != nil {
+			vq.Error("ValidationQueue: block validation failed", log.BlockId(uint64(block.ID())), log.Err(err))
+			vq.updateDependencies(id, false)
+			return
```
Comment (zalmen) [inline sync/validation_queue.go]: Don't you mean `continue`?

**Review by zalmen — CHANGES_REQUESTED** [review]

File: sync/validation_queue.go (hunk 4 above)
Comment (almogdepaz) [reply sync/validation_queue.go]: why yes i do!

File: sync/validation_queue.go (hunk 3 above)
Comment (almogdepaz) [reply sync/validation_queue.go]: removed

**Review by barakshani — APPROVED** [review]

**Comment by almogdepaz** [comment]
resolves   #1401

**Pull request merged**